File: labs/calculator/calculator.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


long add(long a, long b) {
    return a + b;
}

long sub(long a, long b) {
    return a - b;
}

long mult(long a, long b) {
    return a * b;
}


// adds/subtracts/multiplies all values that are in the *values array.
// nValues is the number of values you're reading from the array
// operator will indicate if it's an addition (1), subtraction (2) or
// multiplication (3)
long calc(int operator, int nValues, int *values) {
    long result = values[0];
    long (*opFunc)(long, long);

    switch (operator)
    {
    case 1:
        opFunc = *add;
        break;
    case 2:
        opFunc = *sub;
        break;
    default:
        opFunc = *mult;
        break;
    }

    int currNum;
    for (int i=1; i<nValues; i++) {
        currNum = values[i];
        result = opFunc(result, currNum);
    }

    return result;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        printf("A valid input was not provided.\n");
        printf("Usage example `./calculator.c add 1 2 3 4 5`\n");
        return -1;
    }

    char *op = argv[1];
    if (!(strcmp(op, "add") == 0 || strcmp(op, "sub") == 0 || strcmp(op, "mult") == 0)) {
        printf("A valid operator was not provided.\n");
        printf("Available operators: [add | sub | mult]\n");
        return -1;
    }

    int currNum,
        nums [argc -2];
    char *end;

    for (int i = 2; i < argc; i++) {
        currNum = strtol(argv[i], &end, 10);

        if (*end != '\0') {
            printf("The calculator can only work with integer values!\n");
            printf("Usage example `./calculator.c add 1 2 3 4 5`\n");
            return -1;
        }

        nums[i-2] = currNum;
    }

    int operator_code = 0;

    if (strcmp(op, "add") == 0) operator_code = 1;
    else if (strcmp(op, "sub") == 0) operator_code = 2;
    else operator_code = 3;

    long result = calc(operator_code, argc-2, nums);
    printf("%ld\n", result);

    return 0;
}

File: labs/go-calculator/calculator.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func add(a int, b int) int {
	return a + b
}

func sub(a int, b int) int {
	return a - b
}

func mult(a int, b int) int {
	return a * b
}

func getFuncOp(operator int) func(int, int) int {
	var opFunc func(int, int) int

	if operator ==1 {
        opFunc = add;
	} else if operator == 2 {
        opFunc = sub;
	} else {
        opFunc = mult;
	}

	return opFunc
}

// adds/subtracts/multiplies all values that are in the *values array.
// nValues is the number of values you're reading from the array
// operator will indicate if it's an addition (1), subtraction (2) or
// multiplication (3)
func calc(operator int, values []int) int {
	result := values[0]
	opFunc := getFuncOp(operator)

	for i := 1; i<len(values); i++ {
		result = opFunc(result, values[i])
	}

	return result
}


func main() {
	N := len(os.Args)

	if N < 3 {
		fmt.Printf("A valid input was not provided.\n")
        fmt.Printf("Usage example `./calculator.c add 1 2 3 4 5`\n")
        os.Exit(1)
	}

	op := os.Args[1]
	if !(op == "add" || op == "sub" || op == "mult") {
		fmt.Printf("A valid operator was not provided.\n")
        fmt.Printf("Available operators: [add | sub | mult]\n")
        os.Exit(1)
	}

	nums := make([]int, N - 2)

	for i := 2; i < N; i++ {
		currNum, err := strconv.ParseInt(os.Args[i], 10, 0)
		if err != nil {
			fmt.Printf("The calculator can only work with integer values!\n")
			fmt.Printf("Usage example `./calculator.c add 1 2 3 4 5`\n")
			os.Exit(1)
		}
		nums[i-2] = int(currNum)
	}

	operatorCode := 0

	if op == "add" {
		operatorCode = 1
	} else if op == "sub" {
		operatorCode = 2
	} else {
		operatorCode = 3
	}

	result := calc(operatorCode, nums)
	fmt.Printf(strconv.Itoa(result) + "\n")
}

